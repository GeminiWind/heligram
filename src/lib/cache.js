import { promisify } from 'util';
import redis from 'redis';

const client = redis.createClient('redis://cache');

const get = promisify(client.get).bind(client);
const set = promisify(client.set).bind(client);
const hset = promisify(client.hset).bind(client);
const hget = promisify(client.hget).bind(client);
const del = promisify(client.del).bind(client);
const hdel = promisify(client.hdel).bind(client);

export default {
  get,
  set,
  hset,
  hget,
  del,
  hdel,
};
