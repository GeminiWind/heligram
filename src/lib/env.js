import { config } from 'dotenv';

export default function loadEnv() {
  const envFilePath = process.env.NODE_ENV ? `env.${process.env.NODE_ENV}` : '.env.development';

  config({ path: envFilePath });
}
