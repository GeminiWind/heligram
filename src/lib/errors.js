import JsonApiError from './jsonApiError';

class InternalError extends JsonApiError {
  constructor(message) {
    let internalError = {
      title: 'InternalError',
      code: 'InternalError',
      status: 500,
      detail: 'InternalError',
    };
    if (typeof message === 'string') {
      internalError = {
        ...internalError,
        detail: message,
      };
    }
    if (typeof message === 'object') {
      internalError = {
        ...internalError,
        ...message,
      };
    }
    super(internalError);
  }
}

class NotFoundError extends JsonApiError {
  constructor(message) {
    let notFoundError = {
      title: 'NotFoundError',
      code: 'NotFoundError',
      status: 404,
      detail: 'NotFoundError',
    };
    if (typeof message === 'string') {
      notFoundError = {
        ...notFoundError,
        detail: message,
      };
    }
    if (typeof message === 'object') {
      notFoundError = {
        ...notFoundError,
        ...message,
      };
    }
    super(notFoundError);
  }
}

class BadRequestError extends JsonApiError {
  constructor(message) {
    let badRequestError = {
      title: 'BadRequestError',
      code: 'BadRequestError',
      status: 400,
      detail: 'BadRequestError',
    };
    if (typeof message === 'string') {
      badRequestError = {
        ...badRequestError,
        detail: message,
      };
    }
    if (typeof message === 'object') {
      badRequestError = {
        ...badRequestError,
        ...message,
      };
    }
    super(badRequestError);
  }
}

class MalformedError extends JsonApiError {
  constructor() {
    const malformedError = {
      title: 'MalformedError',
      code: 'MalformedError',
      status: 400,
      detail: 'Request body is malformed.',
    };
    super(malformedError);
  }
}

export {
  InternalError,
  NotFoundError,
  BadRequestError,
  MalformedError,
};
