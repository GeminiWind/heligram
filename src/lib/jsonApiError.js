class JsonApiError extends Error {
  constructor(error) {
    const errorMessage = `${error.title} ${error.detail}`;
    super(errorMessage);

    this.title = error.title;
    this.detail = error.detail;
    this.status = error.status;
    this.code = error.code;

    if (error.source) {
      this.source = error.source;
    }
    if (error.meta) {
      this.meta = error.meta;
    }

    // Ensure the name of this error is the same as the class name
    this.name = this.constructor.name;
    // This clips the constructor invocation from the stack trace.
    // It's not absolutely essential, but it does make the stack trace a little nicer.
    //  @see Node.js reference (bottom)
    Error.captureStackTrace(this, this.constructor);
  }

  toJSON() {
    return {
      status: this.status,
      code: this.code,
      title: this.title,
      detail: this.detail,
      ...(this.source ? { source: this.source } : {}),
      ...(this.meta ? { source: this.meta } : {}),
    };
  }
}

export default JsonApiError;
