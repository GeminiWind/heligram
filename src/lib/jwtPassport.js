import { Strategy as JwtStrategy, ExtractJwt } from 'passport-jwt';
import storageLibrary from './storageLibrary';

export default function jwtPassport(passport) {
  const options = {
    jwtFromRequest: ExtractJwt.fromAuthHeaderAsBearerToken(),
    secretOrKey: process.env.APP_KEY,
  };

  passport.use(new JwtStrategy(options, (jwtPayload, done) => {
    const {
      email,
      scopes,
    } = jwtPayload;

    storageLibrary.findOne({
      Path: `user/${email}`,
    }).then((record) => {
      if (record) {
        const {
          Content: {
            password,
            ...userAttributesWithoutPassword
          },
        } = record;
        done(null, {
          ...userAttributesWithoutPassword,
          scopes,
        });
      } else done(null, false);
    }).catch(error => done(error, false));
  }));
}
