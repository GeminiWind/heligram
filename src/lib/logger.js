import moment from 'moment';
import path from 'path';
import { format, transports, createLogger } from 'winston';

const logger = createLogger({
  format: format.combine(
    format.splat(),
    format.timestamp({
      format: 'YYYY-MM-DD HH:mm:ss',
    }),
    format.printf(info => `${info.timestamp} ${info.level}: ${info.message}`),
  ),
  transports: [
    new transports.Console({
      format: format.combine(
        format.colorize(),
        format.timestamp({
          format: 'YYYY-MM-DD HH:mm:ss',
        }),
        format.printf(info => `${info.timestamp} ${info.level}: ${info.message}`),
      ),
    }),
    new transports.File({
      filename: path.resolve(__dirname, '../..', 'log', `${moment().format('YYYY-MM-DD')}.log`),
    }),
  ],
});

export default logger;
