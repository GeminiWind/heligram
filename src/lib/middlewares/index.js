import loggingHttpRequest from './loggingHttpRequest';
import malformedErrorHandler from './malformedErrorHandler';
import unauthorizedErrorHandler from './unauthorizedErrorHandler';
import jwtAuthz from './jwtAuthz';

export {
  loggingHttpRequest,
  malformedErrorHandler,
  unauthorizedErrorHandler,
  jwtAuthz,
};
