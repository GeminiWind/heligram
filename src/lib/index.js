import logger from './logger';
import storageLibrary from './storageLibrary';

export { logger, storageLibrary };
