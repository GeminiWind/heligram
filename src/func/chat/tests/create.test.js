import {
  validateRequest, createChat, returnResponse,
} from '../create';
import { schemaValidator } from '../../../lib';
import { BadRequestError, InternalError } from '../../../lib/errors';

describe('Create Chat Controller', () => {
  it('can validate request in case that request is matched with schema', () => {
    const request = {
      body: {
        data: {
          type: 'chat',
          attributes: {
            createdBy: 'haidv7@fsoft.com.vn',
            participants: ['haidv7@fsoft.com.vn', 'fake@gmail.com'],
          },
          relationships: {
            messages: {
              data: [
                {
                  type: 'message',
                  attributes: {
                    createdBy: 'haidv7@fsoft.com.vn',
                    message: 'Hello World',
                  },
                },
              ],
            },
          },
        },
      },
      instrumentation: {
        error: jest.fn(),
      },
      schemaValidator,
    };

    let error;

    try {
      validateRequest(request);
    } catch (e) {
      error = e;
    }

    expect(error).toBeFalsy();
  });

  it('throw \'BadRequestError\'  in case that request does not matches with schema', () => {
    const request = {
      body: {
        data: {
          type: 'wrongType',
          attributes: {
            createdBy: 'haidv7@fsoft.com.vn',
            participants: ['haidv7@fsoft.com.vn', 'fake@gmail.com'],
          },
        },
      },
      instrumentation: {
        error: jest.fn(),
      },
      schemaValidator,
    };

    let error;

    try {
      validateRequest(request);
    } catch (e) {
      error = e;
    }

    expect(error).toBeInstanceOf(BadRequestError);
    expect(error).toMatchSnapshot();
  });

  it('can create chat successfuly', async () => {
    const request = {
      body: {
        data: {
          type: 'chat',
          attributes: {
            createdBy: 'haidv7@fsoft.com.vn',
            participants: ['haidv7@fsoft.com.vn', 'fake@gmail.com'],
          },
          relationships: {
            messages: {
              data: [
                {
                  type: 'message',
                  attributes: {
                    createdBy: 'haidv7@fsoft.com.vn',
                    message: 'Hello World',
                  },
                },
              ],
            },
          },
        },
      },
      storageLibrary: {
        create: jest.fn().mockImplementation(() => Promise.resolve(true)),
      },
    };

    const state = await createChat(request);

    expect(state).toHaveProperty('chatId');
  });

  it('throw \'InternalError\'  in case that error in saving chat into storage', async () => {
    const request = {
      body: {
        data: {
          type: 'chat',
          attributes: {
            createdBy: 'haidv7@fsoft.com.vn',
            participants: ['haidv7@fsoft.com.vn', 'fake@gmail.com'],
          },
          relationships: {
            messages: {
              data: [
                {
                  type: 'message',
                  attributes: {
                    createdBy: 'haidv7@fsoft.com.vn',
                    message: 'Hello World',
                  },
                },
              ],
            },
          },
        },
      },
      storageLibrary: {
        create: jest.fn().mockImplementation(() => Promise.reject(new Error('Runtime error.'))),
      },
      instrumentation: {
        error: jest.fn(),
      },
    };

    let error;

    try {
      await createChat(request);
    } catch (e) {
      error = e;
    }

    expect(error).toBeInstanceOf(InternalError);
    expect(error).toMatchSnapshot();
  });

  it('can return response correctly', () => {
    const request = {
      body: {
        data: {
          type: 'chat',
          attributes: {
            createdBy: 'haidv7@fsoft.com.vn',
            participants: ['haidv7@fsoft.com.vn', 'fake@gmail.com'],
          },
          relationships: {
            messages: {
              data: [
                {
                  type: 'message',
                  attributes: {
                    createdBy: 'haidv7@fsoft.com.vn',
                    message: 'Hello World',
                  },
                },
              ],
            },
          },
        },
      },
      storageLibrary: {
        create: jest.fn().mockImplementation(() => Promise.resolve(new Error('Runtime error.'))),
      },
      chatId: 'fake-id',
    };

    const response = returnResponse(request);

    expect(response).toHaveProperty('statusCode', 201);
    expect(response).toHaveProperty('body');
    expect(response).toMatchSnapshot();
  });
});
