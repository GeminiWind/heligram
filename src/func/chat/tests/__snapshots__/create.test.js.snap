// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Create Chat Controller can return response correctly 1`] = `
Object {
  "body": Object {
    "data": Object {
      "attributes": Object {
        "createdBy": "haidv7@fsoft.com.vn",
        "participants": Array [
          "haidv7@fsoft.com.vn",
          "fake@gmail.com",
        ],
      },
      "id": "fake-id",
      "relationships": Object {
        "messages": Object {
          "data": Array [
            Object {
              "attributes": Object {
                "createdBy": "haidv7@fsoft.com.vn",
                "message": "Hello World",
              },
              "type": "message",
            },
          ],
        },
      },
      "type": "chat",
    },
  },
  "statusCode": 201,
}
`;

exports[`Create Chat Controller throw 'BadRequestError'  in case that request does not matches with schema 1`] = `[BadRequestError: BadRequestError BadRequestError]`;

exports[`Create Chat Controller throw 'InternalError'  in case that error in saving chat into storage 1`] = `[InternalError: InternalError Error in creating chat in storage.]`;
